boards: arm64: Add support for Intel SoC FPGA Agilex5 platform.

gdengi opened the pull request:
The Intel Agilex® 5 FPGA product family extends the innovations of the Intel Agilex® FPGA portfolio to midrange FPGA applications.
Added support for Intel SoC FPGA Agilex5 platform in the Zephyr with the following sub-modules and updates.

- Add 'intel_socfpga_agilex5_socdk' board
- Add 'agilex5' soc to Intel Agilex soc series
- Reset manager for Intel SoC FPGA Agilex5 platform
- Clock manager for Intel SoC FPGA Agilex5 platform
- UART, add an optional module reset line functionality
- PSCI, add support for multiple PSCI versions 0.2v and 1.1v
- Custom shell module configurations added as required by the Intel Agilex platforms.

gdengi committed:
soc: arm64: Add agilex5 soc folder and its configurations

Add Agilex5 soc folder, MMU table and its configurations for
Intel SoC FPGA Agilex5 platform for initial bring up.
Add ARM Cortex-a76 and Cortex-a55 HMP cluster type.

Signed-off-by: Teik Heng Chong <teik.heng.chong@intel.com>
Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>

gdengi committed:
boards: arm64: Add Intel SoC FPGA Agilex5 development kit board

This is the initial Zephyr support for Intel SoC FPGA Agilex5 support.
Agilex5 has dual-core 64-bit ARM Cortex*-A55 and dual-core 64bit
ARM Cortex*-A76.

The Zephyr will need to be loaded by Intel Arm Trusted Firmware (ATF).
Agilex5 Zephyr boot flow:
  FSBL:ATF BL2(EL3) -> ATF BL31(EL3) -> OS:Zephyr(EL1)

Intel ATF can be loaded from:
  https://github.com/altera-opensource/arm-trusted-firmware.git

Signed-off-by: Teik Heng Chong <teik.heng.chong@intel.com>
Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>

gdengi committed:
drivers: clock_control: clock driver for Intel Agilex5 platform

This is Intel's proprietary IP which supply the clock for all the
system peripherals. Clock manager is initialized only one time
during boot up by FSBL (ATF BL2) based on external user settings.

Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>

gdengi committed:
dts: arm64: Add device tree for Intel SoCFPGA Agilex5 platform

Device tree for Intel SoCFPGA Agilex5 initial bring up. This is the
first version of device tree which enable four cores SMP and basic
drivers that needed by 'hello_world' and 'cli' applications.

Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>

gdengi committed:
drivers: reset: Add reset controller for Intel Agilex5 platform

This is Intel's proprietary IP which controls individual module
reset signals. During each system driver initialization, these
reset signals will be used to bring module out of reset state.

Signed-off-by: Navinkumar Balabakthan <navinkumar.balabakthan@intel.com>
Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>

gdengi committed:
drivers: serial: Add optional reset line for uart_ns16550

If the optional hardware reset line is available, this change
will use that reset line to assert the uart module and bring
it out of reset state to use.

Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>

gdengi committed:
drivers: pm_cpu_ops: Add support for multiple PSCI versions

Each PSCI interface versions have different DT compatible strings
like arm,psci-0.2, arm,psci-1.1 and so on. However, the same driver
can be used for all the versions by adding #define DT_COMPAT for
required version and #undef DT_COMPAT for default version.

Add support for PSCI cold reset, warm reset and cpu-on function IDs.

Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>
Signed-off-by: Navinkumar Balabakthan <navinkumar.balabakthan@intel.com>

gdengi committed:
CODEOWNERS: Add code owners for Intel Agilex5 platform drivers

Code owners for all the required Intel Agilex5 platform drivers.

Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>

povergoing commented on boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:
```diff
@@ -0,0 +1,51 @@
+CONFIG_RESET_INIT_PRIORITY=35
+
+# SOC
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=400000000
```
Can you elaborate on why you set 25000000 in soc as a default value, but here you force it as 400000000

gdengi replied:
Its a mistake. Removed from here, and added 400000000 in soc as a default value.

povergoing commented on boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:
```diff
@@ -0,0 +1,51 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# The Zephyr build from this defconfig is execpted to boot from
+#Intel Arm Trusted Firmware (ATF).
```
What is the `Intel Arm Trusted Firmware`, it's an Intel fork of ATF or ?

gdengi replied:
Yes, it is - where we have support for Intel SoC FPGA product lines.
And we are using the ATF as a boot loader to Zephyr.

povergoing commented on boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst:
```diff
@@ -0,0 +1,14 @@
+Intel Agilex5 SoC Development Kit
+#################################
+
+Overview
```
It would be better to put more details like other boards did?

gdengi replied:
Will add shortly.

gdengi replied:
Updated with all the details. Thank you.

povergoing commented on drivers/clock_control/clock_control_agilex5.c:
```diff
@@ -0,0 +1,90 @@
+}
+static const struct clock_control_driver_api clock_api = {.get_rate = clock_get_rate};
+
+#define CLOCK_CONTROL_DEVICE(_inst)					\
```
Would you like to fix the alignments here?

gdengi replied:
Updated.

carlocaione reviewed (commented):
In general looks good. Just a couple of comments.

fabiobaltieri commented on boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:
```diff
@@ -0,0 +1,51 @@
+
+# Setting higher priority for Intel SOCFPGA reset controller to initialize it before
+# other dependent drivers running at CONFIG_KERNEL_INIT_PRIORITY_DEFAULT which is 40
+CONFIG_RESET_INIT_PRIORITY=35
```
Ideally... I'll hack it in my test PR and see if anything else complains, hang in there.

fabiobaltieri replied:
Run result: https://github.com/zephyrproject-rtos/zephyr/pull/58388/checks?check_run_id=14941982543, these were all failing already so yes - I think it would be ok to anticipate `RESET_INIT_PRIORITY` to 35 system wide. Maybe do a separate PR for it?

gdengi replied:
So, shall I keep it here as of now as part this commit?

carlocaione replied:
can you open a different PR to fix that?

gdengi replied:
Sure, will do. Thank you

gdengi replied:
@carlocaione  @fabiobaltieri  FYI, created a separate PR for this change. Thank you.
https://github.com/zephyrproject-rtos/zephyr/pull/60412

ickochar commented on drivers/pm_cpu_ops/pm_cpu_ops_psci.c:
```diff
@@ -19,7 +21,11 @@ LOG_MODULE_REGISTER(psci);
 #include "pm_cpu_ops_psci.h"
 
-static struct psci psci_data;
+/* PSCI device configuration object macro. */
+#define DEV_CFG(_dev) ((const struct psci_config_t *)(_dev)->config)
```
Any alternative suggested for such macros ?

carlocaione replied:
just move `const struct config = dev->config` directly in place.

gdengi replied:
Updated to _const struct config = dev->config_ style and removed DEV_CFG macro.

gdengi commented on drivers/pm_cpu_ops/pm_cpu_ops_psci.h:
```diff
@@ -39,6 +39,65 @@
+#define PSCI_1_1_64BIT	   (0x40000000U)
+#define PSCI_1_1_FN64_BASE (PSCI_1_1_FN_BASE + PSCI_1_1_64BIT)
+#define PSCI_1_1_FN(n)	   (PSCI_1_1_FN_BASE + (n))
+#define PSCI_1_1_FN64(n)   (PSCI_1_1_FN64_BASE + (n))
```
Updated.

dcpleung commented on drivers/serial/uart_ns16550.c:
```diff
@@ -229,6 +233,8 @@ struct uart_ns16550_device_config {
 	bool io_map;
 #endif
+
+	struct reset_dt_spec reset_spec;
```
Please use `DT_ANY_INST_HAS_PROP_STATUS_OKAY` to enable reset related stuff. If there is no reset controller attached (which is the majority of cases), there is no need to include all the added code and data.

gdengi replied:
Updated the driver with DT_ANY_INST_HAS_PROP_STATUS_OKAY. Thank you.

dcpleung reviewed (changes requested):
Also, could you use `samples/subsys/shell/shell_module/` instead of creating a new sample? The `shell_module` should also provide a shell which you can use to query devices.

gdengi commented:
> Also, could you use `samples/subsys/shell/shell_module/` instead of creating a new sample? The `shell_module` should also provide a shell which you can use to query devices.

We wanted to keep simple and light CLI application for Intel SoC FPGA products line and that's where we introduce this application.
Sure, will try out the suggested sample and test. Will keep you updated. Thank you.

gdengi commented on boards/arm64/intel_socfpga_agilex5_socdk/intel_socfpga_agilex5_socdk_defconfig:
```diff
@@ -0,0 +1,44 @@
+# Copyright (c) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+# The Zephyr build from this defconfig is execpted to boot from
```
Corrected.

gdengi committed:
samples: subsys: shell: Custom configs for intel_socfpga_agilex* boards

The intent of this change is to add custom shell configurations for
intel_socfpga_agilex* based boards. As of now, configurations are
added for 'intel_socfpga_agilex5_socdk' board.

Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>

gdengi commented:
Hello @carlocaione  

All the review comments are addressed and all the required CI checks are passed.
Kindly let me know your inputs on the further steps.

Thank you.

Regards,
Girisha Dengi

fabiobaltieri commented on drivers/serial/uart_ns16550.c:
```diff
@@ -576,6 +590,34 @@ static int uart_ns16550_config_get(const struct device *dev,
+ */
+static int uart_reset_config(const struct reset_dt_spec *reset_spec)
+{
+	int ret = 0;
```
Drop the initializer.

gdengi replied:
Updated

fabiobaltieri commented on drivers/pm_cpu_ops/psci_shell.c:
```diff
@@ -0,0 +1,73 @@
+	ARG_UNUSED(shctx);
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+	int ret = 0;
```
drop the initializer

gdengi replied:
Updated

fabiobaltieri commented on drivers/pm_cpu_ops/psci_shell.c:
```diff
@@ -0,0 +1,73 @@
+	ARG_UNUSED(shctx);
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+	int ret = 0;
```
same

gdengi replied:
Updated

fabiobaltieri commented on drivers/pm_cpu_ops/psci_shell.c:
```diff
@@ -0,0 +1,73 @@
+{
+	ARG_UNUSED(shctx);
+	ARG_UNUSED(argc);
+	long cpu_id = 0;
+	int result = 0;
```
same

gdengi replied:
Updated

fabiobaltieri commented on CODEOWNERS:
Does `MAINTAINERS.yml` needs to be updated as well?

gdengi replied:
Updated

ickochar commented on soc/arm64/intel_socfpga/agilex5/Kconfig.defconfig.agilex5:
```diff
@@ -0,0 +1,21 @@
+# - include the UART interrupts 173 or 204
+config NUM_IRQS
+	int
+	default 205
```
update to value above 242

gdengi replied:
Updated now as minimal required for initial Agilex5 BSP. 
Will update this as and when required drivers are implemented.

gdengi committed:
MAINTAINERS: Add self as maintainer for Intel Agilex platform

Add maintainer and collaborators for Intel Agilex platforms and
related drivers.

Signed-off-by: Girisha Dengi <girisha.dengi@intel.com>

gdengi commented on drivers/pm_cpu_ops/pm_cpu_ops_psci.c:
```diff
@@ -80,6 +83,48 @@ int pm_system_off(void)
+	int result = 0;
+
+	/* PSCI_FEATURES function ID is supported from PSCI 1.0 onwards. */
+	if (PSCI_VERSION_MAJOR(psci_data.ver) >= 1) {
```
Updated as suggested. Thank you

gdengi commented on drivers/serial/uart_ns16550.c:
```diff
@@ -53,6 +56,14 @@ BUILD_ASSERT(IS_ENABLED(CONFIG_PCIE), "NS16550(s) in DT need CONFIG_PCIE");
 
+/* Is UART module 'resets' line property defined */
+#define UART_NS16550_RESET_ENABLED \
+	DT_ANY_INST_HAS_PROP_STATUS_OKAY(resets)
```
Updated

gdengi replied:
Updated

gdengi commented on drivers/serial/uart_ns16550.c:
```diff
@@ -576,6 +590,34 @@ static int uart_ns16550_config_get(const struct device *dev,
+
+	ret = reset_line_toggle(reset_spec->dev, reset_spec->id);
+	if (ret != 0) {
+		LOG_ERR("UART toggle reset line failed");
```
Updated

gdengi commented on drivers/pm_cpu_ops/pm_cpu_ops_psci.c:
```diff
@@ -80,6 +83,48 @@ int pm_system_off(void)
+ */
+static int psci_features_check(unsigned long function_id)
+{
+	int result = 0;
```
Updated

gdengi commented on dts/arm64/intel/intel_socfpga_agilex5.dtsi:
```diff
@@ -0,0 +1,124 @@
+	arch_timer: timer {
+		compatible = "arm,armv8-timer";
+		interrupt-parent = <&gic>;
+		interrupts = <GIC_PPI 13 IRQ_TYPE_LEVEL
```
Updated

gdengi commented on drivers/clock_control/clock_control_agilex5_ll.c:
```diff
@@ -0,0 +1,181 @@
+/* Calculate clock frequency to be used for mpu */
+uint32_t get_mpu_clk(void)
+{
+	uint32_t mpu_clk = 0;
```
Updated

gdengi commented on include/zephyr/dt-bindings/reset/intel_socfpga_reset.h:
```diff
@@ -0,0 +1,135 @@
+
+
+#define		RSTMGR_SDMCOLDRST_RSTLINE			0
+#define		RSTMGR_SDMWARMRST_RSTLINE			1
```
Updated

gdengi commented on drivers/reset/reset_intel_socfpga.c:
```diff
@@ -0,0 +1,115 @@
+	const struct reset_intel_config *config = (const struct reset_intel_config *)dev->config;
+	uintptr_t base_address = DEVICE_MMIO_GET(dev);
+	uint16_t offset = 0;
+	uint8_t regbit = 0;
```
Updated

gdengi commented on drivers/reset/reset_intel_socfpga.c:
```diff
@@ -0,0 +1,115 @@
+	uintptr_t base_address = DEVICE_MMIO_GET(dev);
+	uint32_t value = 0;
+	uint16_t offset = 0;
+	uint8_t regbit = 0;
```
Updated

gdengi commented:
@fabiobaltieri Addressed all the review comments. Kindly take a look. Thank you.

gdengi commented on boards/arm64/intel_socfpga_agilex5_socdk/doc/index.rst:
```diff
@@ -0,0 +1,80 @@
+provides services like SMC (Secure monitor calls) and PSCI (Power state coordination interface).
+
+Boot flow:
+        ATF BL2 (EL3) -> ATF BL31 (EL3) -> Zephyr (EL2->EL1)
```
Updated

ickochar reviewed (approved):
Looks good to me!

gdengi commented:
Hello @carlocaione, 

All the required approvals are in place - kindly take a look and help us to merge.
Thank you.

carlocaione commented:
> All the required approvals are in place - kindly take a look and help us to merge. Thank you.

Unfortunately I don't have that power. But the mighty @fabiobaltieri has it.

gdengi commented:
@dcpleung Kindly review and provide your feedback. Thank you.

gdengi commented:
> > All the required approvals are in place - kindly take a look and help us to merge. Thank you.
> 
> Unfortunately I don't have that power. But the mighty @fabiobaltieri has it.
> 
> Also @dcpleung needs to review this again.

Thank you @carlocaione for the info.

gdengi commented:
@fabiobaltieri  All the review comments are addressed and all the approvals are in place.
Kindly take a look and help us to merge.

Thank you.

stephanosio commented on MAINTAINERS.yml:
```diff
@@ -2288,6 +2291,22 @@ Intel Platforms (Xtensa):
+    - gdengi
+  collaborators:
+    - nbalabak
+    - chongteikheng
```
@gdengi Hi. The collaborator added here, `chongteikheng`, does not seem to be a valid GitHub account. Please submit a patch to either update or remove the username from the MAINTAINERS file.

gdengi replied:
Hello @stephanosio, sure will submit a patch to update the account name here. Thank you.

gdengi replied:
Hello @stephanosio, FYI, created a PR for the same. https://github.com/zephyrproject-rtos/zephyr/pull/65190

stephanosio commented on CODEOWNERS:
```diff
@@ -201,6 +201,7 @@
 /boards/arm64/fvp_baser_aemv8r/           @povergoing
 /boards/arm64/fvp_base_revc_2xaemv8a/     @carlocaione
 /boards/arm64/intel_socfpga_agilex_socdk/ @siclim @ngboonkhai
+/boards/arm64/intel_socfpga_agilex5_socdk/ @chongteikheng
```
ditto

gdengi replied:
Sure, will update here also.

gdengi replied:
Created a PR for the same. https://github.com/zephyrproject-rtos/zephyr/pull/65190
